### PR TITLE
ddl: restrain drop column with index using the same switch as EnableChangeMultiSchema

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: ailinkid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: restrain drop column with index using the same switch as EnableChangeMultiSchema

### What is changed and how it works?

when drop column with index on it, check it.
passing the front-end switch in job args to the ddl worker

### Check List 

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)

### Release note 

```release-note
ddl: restrain drop column with index using the same switch as EnableChangeMultiSchema
```

### CODE REVIEW

File: ddl/column.go [hunk 1]
```diff
@@ -507,7 +507,8 @@ func checkDropColumns(t *meta.Meta, job *model.Job) (*model.TableInfo, []*model.
 
 	var colNames []model.CIStr
 	var ifExists []bool
-	err = job.DecodeArgs(&colNames, &ifExists)
+	var allowDropColumnWithIndex bool
+	err = job.DecodeArgs(&colNames, &ifExists, &allowDropColumnWithIndex)
```
Comment (zimulala) [inline ddl/column.go:511]: Could we put it to `MultiSchemaInfo`?

File: go.mod [hunk 2]
```diff
@@ -99,3 +99,5 @@ require (
 
 // FIXME the official repo has some bug makes br_gcs test failed. https://github.com/googleapis/google-cloud-go/pull/3509
 // replace cloud.google.com/go/storage => github.com/3pointer/google-cloud-go/storage v1.6.1-0.20210108125931-b59bfa0720b2
+
+replace github.com/pingcap/parser => github.com/ailinkid/parser v0.0.0-20210823091130-bfdf9753fa15
```
Comment (bb7133) [inline go.mod:103]: What is the related parser PR?

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -3376,12 +3376,13 @@ func (d *ddl) DropColumn(ctx sessionctx.Context, ti ast.Ident, spec *ast.AlterTa
...
-		SchemaName: schema.Name.L,
-		Type:       model.ActionDropColumn,
-		BinlogInfo: &model.HistoryInfo{},
-		Args:       []interface{}{colName},
+		SchemaID:        schema.ID,
+		TableID:         t.Meta().ID,
+		SchemaName:      schema.Name.L,
+		Type:            model.ActionDropColumn,
+		BinlogInfo:      &model.HistoryInfo{},
+		MultiSchemaInfo: &model.MultiSchemaInfo{},
```
Comment (tangenta) [inline ddl/ddl_api.go:3384]: We should check the system variable `tidb_enable_change_multi_schema` to initialize this field.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -3448,12 +3449,13 @@ func (d *ddl) DropColumns(ctx sessionctx.Context, ti ast.Ident, specs []*ast.Alt
...
-		SchemaName: schema.Name.L,
-		Type:       model.ActionDropColumns,
-		BinlogInfo: &model.HistoryInfo{},
-		Args:       []interface{}{colNames, ifExists},
+		SchemaID:        schema.ID,
+		TableID:         t.Meta().ID,
+		SchemaName:      schema.Name.L,
+		Type:            model.ActionDropColumns,
+		BinlogInfo:      &model.HistoryInfo{},
+		MultiSchemaInfo: &model.MultiSchemaInfo{},
```
Comment (tangenta) [inline ddl/ddl_api.go:3457]: Ditto.

**Review by tangenta — COMMENTED** [review]
Could you please add a simple test?
```sql
drop table if exists t;
create table t (a int, b int, index(a));
set @@global.tidb_enable_change_multi_schema = false;
alter table t drop column a;  -- failed
set @@global.tidb_enable_change_multi_schema = true;
alter table t drop column a;  -- succeed
```

**Review by zimulala — COMMENTED** [review]
Please add tests.

File: ddl/column.go [hunk 5]
```diff
@@ -1733,9 +1733,19 @@ func isColumnWithIndex(colName string, indices []*model.IndexInfo) bool {
...
+			return len(idx.Columns) > 1
+		}
+	} else {
+		condition = func(idx *model.IndexInfo) bool {
+			return len(idx.Columns) >= 1
+		}
+	}
 	for _, indexInfo := range indices {
-		if indexInfo.Primary || len(indexInfo.Columns) > 1 {
+		if indexInfo.Primary || condition(indexInfo) {
```
Comment (tangenta) [inline ddl/column.go:1748]: How about 
```go
if indexInfo.Primary || len(indexInfo.Columns) > 1) || 
(!enabledMultiSchemaChange && len(indexInfo.Columns) == 1) {
```
and remove those anonymous functions?

Comment (AilinKid) [reply ddl/column.go:1748]: good

**Review by tangenta — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**